gk/main.c

```diff
@@ -478,6 +479,75 @@ get_block_idx(struct gk_config *gk_conf, unsigned int lcore_id)
 	return 0;
 }
 
+static bool
+is_flow_expired(struct flow_entry *fe,
+	uint64_t now, uint64_t request_timeout_cycles)
+{
+	switch(fe->state) {
+	case GK_REQUEST:
+		if (fe->u.request.last_packet_seen_at > now) {
+			char err_msg[128];
+			int ret = snprintf(err_msg, sizeof(err_msg),
+				"gk: buggy condition at %s wrong timestamp",
+				__func__);
+			RTE_VERIFY(ret > 0 && ret < (int)sizeof(err_msg));
+			print_flow_err_msg(&fe->flow, err_msg);
+			return true;
+		}
+
+		return now - fe->u.request.last_packet_seen_at >=
+			request_timeout_cycles;
+	case GK_GRANTED:
+		return now >= fe->u.granted.cap_expire_at;
+	case GK_DECLINED:
+		return now >= fe->u.declined.expire_at;
+	default:
+		return true;
+	}
+}
+
+static int
+gk_del_flow_entry_from_hash(struct rte_hash *h, struct flow_entry *fe)
+{
+	int ret = rte_hash_del_key(h, &fe->flow);
+	if (likely(ret >= 0))
+		memset(fe, 0, sizeof(*fe));
+	else if (ret == -EINVAL) {
+		RTE_LOG(ERR, HASH,
+			"gk: The parameters are invalid at %s",
+			__func__);
+	} else if (ret == -ENOENT) {
+		RTE_LOG(ERR, HASH,
+			"gk: The flow key is not found at %s",
+			__func__);
+	}
+
+	return ret;
+}
+
+static void
+gk_flow_tbl_bucket_scan(uint32_t *bidx, uint32_t *next,
+	uint64_t request_timeout_cycles, struct gk_instance *instance)
```

Comment:
`Parameter `next` should be a local variable since we always want to scan a whole bucket. It simplifies the code a little.